tests/lib/utils/selector.js

```diff
@@ -0,0 +1,89 @@
+'use strict'
+
+const fs = require('fs')
+const path = require('path')
+const assert = require('assert')
+
+const Linter = require('eslint').Linter
+
+const selector = require('../../../lib/utils/selector')
+const utils = require('../../../lib/utils')
+
+const FIXTURE_ROOT = path.resolve(__dirname, '../../fixtures/utils/selector')
+
+/**
+ * Load test patterns from fixtures.
+ *
+ * @returns {object} The loaded patterns.
+ */
+function loadPatterns() {
+  return fs.readdirSync(FIXTURE_ROOT).map((name) => {
+    const code0 = fs.readFileSync(
+      path.join(FIXTURE_ROOT, name, 'source.vue'),
+      'utf8'
+    )
+    const code = code0.replace(/^<!--(.+?)-->/, `<!--${name}-->`)
+    const inputSelector = /^<!--(.+?)-->/.exec(code0)[1].trim()
+    return { code, name, inputSelector }
+  })
+}
+
+function extractElements(code, inputSelector) {
+  const linter = new Linter()
+  const result = []
+
+  linter.defineRule('vue/selector-test', (context) => {
+    const parsed = selector.parseSelector(inputSelector, context)
+    return utils.defineDocumentVisitor(context, {
+      VElement(node) {
+        if (parsed.test(node)) {
+          result.push({
+            text: context.getSourceCode().text.slice(...node.startTag.range),
+            elementText: parsed.elementText
+          })
```

Comment:
`elementText` is never used/printed anywhere else. Is that intended? Otherwise, this might be simplified to:

```suggestion
          result.push(context.getSourceCode().text.slice(...node.startTag.range))
```